### PR TITLE
Implement MintX509SVID on API v2

### PR DESCRIPTION
**Pull request by MarcosDY**
Implements server.MintX509SVID and service.MintX509SVID

Solve: #1527

### CODE REVIEW

File: pkg/server/api/svid/v1/server.go [hunk 1]
```diff
@@ -18,8 +22,40 @@ type server struct {
...
+		log.WithError(err).Error("Invalid CSR")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: %v", err)
+	}
+
+	x509Svid, err := s.service.MintX509SVID(ctx, csr, time.Duration(req.Ttl)*time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	svidChain := make([][]byte, 0, len(x509Svid.CertChain))
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server.go:44]: We should make a method for this under the x509util package.

File: pkg/server/api/svid/v1/server_test.go [hunk 2]
```diff
@@ -2,35 +2,91 @@ package svid_test
...
+func TestMintX509SVID(t *testing.T) {
+	log, _ := test.NewNullLogger()
+	ctx := rpccontext.WithLogger(context.Background(), log)
 
-	contextFn := func(ctx context.Context) context.Context {
-		return ctx
+	fakeService := &FakeService{}
+	client := createClient(ctx, t, fakeService)
+
+	spiffeID := spiffeid.Must("trust.domain", "workload1")
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:31]: nit: we should use `example.org` or `<something>.test` domains since those are intended for test use.

File: pkg/server/api/svid/v1/service.go [hunk 3]
```diff
@@ -24,15 +35,73 @@ type Service interface {
...
+
+	spiffeID, err := spiffeid.FromURI(csr.URIs[0])
+	if err != nil {
+		log.WithError(err).Error("Invalid SPIFFE ID")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid SPIFFE ID: %v", err)
+	}
+
+	if !spiffeID.MemberOf(s.TrustDomain) {
+		log.Error("Invalid SPIFFE ID: not member of servers trust domain")
+		return nil, status.Error(codes.InvalidArgument, "invalid SPIFFE ID: not member of servers trust domain")
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:71]:
```suggestion
		log.Error("Invalid SPIFFE ID: not member of the server trust domain")
		return nil, status.Error(codes.InvalidArgument, "invalid SPIFFE ID: not member of the server trust domain")
```

File: pkg/server/api/svid/v1/service.go [hunk 4]
```diff
@@ -24,15 +35,73 @@ type Service interface {
...
+		subject = &csr.Subject
+	}
+
+	svid, err := s.ServerCA.SignX509SVID(ctx, ca.X509SVIDParams{
+		SpiffeID:  spiffeID.String(),
+		PublicKey: csr.PublicKey,
+		// SignX509SVID is taking care of of ttl comparation against bundle
+		TTL:     ttl,
+		DNSList: csr.DNSNames,
+		Subject: subject,
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:92]: I think this is a little more ergonomic if `Subject` is not a pointer. Then this code can just set unconditionally pass `csr.Subject`. The server CA code can check if `Subject.String() == ""` to determine if it needs to apply the default.

File: pkg/server/api/svid/v1/service.go [hunk 5]
```diff
@@ -24,15 +35,73 @@ type Service interface {
...
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to sign X.509 SVID")
+		return nil, status.Error(codes.Internal, err.Error())
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:96]:
```suggestion
		return nil, status.Errorf(codes.Internal, "failed to sign X509-SVID: %v", err)
```

File: pkg/server/api/svid/v1/service.go [hunk 6]
```diff
@@ -24,15 +35,73 @@ type Service interface {
...
+	}
+
+	// TODO: may I verify svid is not empty?
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:99]: I think it's safe to assume the chain returned by the server CA will be non-empty.

File: pkg/server/api/svid/v1/service.go [hunk 7]
```diff
@@ -24,15 +35,73 @@ type Service interface {
...
+	return &api.X509SVID{
+		ID:        spiffeID,
+		CertChain: svid,
+		ExpiresAt: svid[0].NotAfter.Local().UTC(),
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:103]: Curious why you convert to `Local()` before converting to `UTC()`? The certificate `notAfter` should already be in UTC, but if we want to be defensive, we can just call `UTC()` without first going to `Local()`...
```suggestion
		ExpiresAt: svid[0].NotAfter.UTC(),
```

File: .go-version [hunk 8]
```diff
@@ -1 +1 @@
-1.13.9
+1.14.2
```
Comment (azdagron) [inline .go-version]: Any moves to a new version of go deserve their own PR. That being said, we intentionally reverted back to go1.13 from go1.14 due to the preemptive scheduler ticking a bug in the linux kernel. We shouldn't move forward until we have determined that the risk of hitting an unpatched kernel is low.

File: pkg/server/api/svid/v1/service.go (hunk 7 above)
Comment (MarcosDY) [reply pkg/server/api/svid/v1/service.go:103]: that was a typo.. my idea was always do just a `UTC`

File: pkg/server/api/svid/v1/server_test.go [hunk 9]
```diff
@@ -2,35 +2,88 @@ package svid_test
...
-	"google.golang.org/grpc/codes"
 )
 
-func TestServer(t *testing.T) {
-	registerFn := func(s *grpc.Server) {
-		svid.RegisterService(s, FakeService{})
-	}
+func TestMintX509SVID(t *testing.T) {
+	log, _ := test.NewNullLogger()
+	ctx := rpccontext.WithLogger(context.Background(), log)
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:25]: I think it's confusing to create this context here, embellish it with things intended for the server, and then use it to make the RPC call with the client. It's also problematic that the request context is being dropped in `contextFn`, we should extend that context and return it instead of dropping it and returning a new one.

I'd suggest that:
- The `createClient` function not take a context, but rather actively embellish the request context via the `contextFn` callback.
- The null logger be created inside of `createClient` unless we want to test that the right stuff is logged, in which case, `createClient` could return the `hook` returned from `test.NewNullLogger`

File: pkg/server/api/svid/v1/server_test.go [hunk 10]
```diff
@@ -2,35 +2,88 @@ package svid_test
...
+	csrRaw, err := x509.CreateCertificateRequest(rand.Reader, template, key)
+	require.NoError(t, err)
 
-	conn, done := spiretest.NewAPIServer(t, registerFn, contextFn)
-	defer done()
+	// Setup fake
+	expiresAt := time.Now()
+	fakeService.svid = &api.X509SVID{
+		ID:        spiffeID,
+		CertChain: []*x509.Certificate{},
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:46]: we should return something here so we can  verify the certificates are added appropriately to the response

File: pkg/server/api/svid/v1/server_test.go [hunk 11]
```diff
@@ -2,35 +2,88 @@ package svid_test
...
+	template := &x509.CertificateRequest{
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+		URIs:               []*url.URL{spiffeID.URL()},
 	}
+	csrRaw, err := x509.CreateCertificateRequest(rand.Reader, template, key)
+	require.NoError(t, err)
 
-	conn, done := spiretest.NewAPIServer(t, registerFn, contextFn)
-	defer done()
+	// Setup fake
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:42]: nit:
```suggestion
	// Set up fake server
```

File: pkg/server/api/svid/v1/server_test.go [hunk 12]
```diff
@@ -2,35 +2,88 @@ package svid_test
...
+		Ttl: 10,
+	})
+	require.NoError(t, err)
 
-	resp, err := client.MintX509SVID(context.Background(), &svidpb.MintX509SVIDRequest{})
-	spiretest.RequireGRPCStatus(t, err, codes.Unimplemented, "not implemented")
-	require.Nil(t, resp)
+	require.Equal(t, "example.org", resp.Svid.Id.TrustDomain)
+	require.Equal(t, "/workload1", resp.Svid.Id.Path)
+	require.Equal(t, expiresAt.UTC().Unix(), resp.Svid.ExpiresAt)
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:59]: We should assert the certificate chain content

File: pkg/server/api/svid/v1/server_test.go [hunk 13]
```diff
@@ -2,35 +2,88 @@ package svid_test
...
+
+func (s *FakeService) MintX509SVID(ctx context.Context, csr *x509.CertificateRequest, ttl time.Duration) (*api.X509SVID, error) {
+	if s.err != nil {
+		return nil, s.err
+	}
+
+	return s.svid, nil
+}
+
+func createClient(ctx context.Context, t *testing.T, fakeService *FakeService) (svidpb.SVIDClient, func()) {
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:77]: nit: `createClient` ==> `startFakeService`

File: pkg/server/api/svid/v1/server_test.go [hunk 14]
```diff
@@ -2,35 +2,88 @@ package svid_test
...
+	registerFn := func(s *grpc.Server) {
+		svid.RegisterService(s, fakeService)
+	}
+
+	contextFn := func(context.Context) context.Context {
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:82]: This callback should return (after optionally embellishing) the passed in context instead of dropping it.

File: pkg/server/api/svid/v1/service.go [hunk 15]
```diff
@@ -24,15 +34,67 @@ type Service interface {
...
+
+	spiffeID, err := spiffeid.FromURI(csr.URIs[0])
+	if err != nil {
+		log.WithError(err).Error("Invalid SPIFFE ID")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid SPIFFE ID: %v", err)
+	}
+
+	if !spiffeID.MemberOf(s.TrustDomain) {
+		log.Error("Invalid SPIFFE ID: not member of the servers trust domain")
+		return nil, status.Error(codes.InvalidArgument, "invalid SPIFFE ID: not member of the servers trust domain")
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:70]:
```suggestion
		log.Error("Invalid SPIFFE ID: not member of the server trust domain")
		return nil, status.Error(codes.InvalidArgument, "invalid SPIFFE ID: not member of the server trust domain")
```

File: pkg/server/api/svid/v1/service.go [hunk 16]
```diff
@@ -24,15 +34,67 @@ type Service interface {
...
+		if err := validateDNS(dnsName); err != nil {
+			log.WithError(err).Error("Invalid DNS name")
+			return nil, status.Errorf(codes.InvalidArgument, "invalid DNS name: %v", err)
+		}
+	}
+
+	svid, err := s.ServerCA.SignX509SVID(ctx, ca.X509SVIDParams{
+		SpiffeID:  spiffeID.String(),
+		PublicKey: csr.PublicKey,
+		// SignX509SVID is taking care of of ttl comparation against bundle
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:83]: Can you elaborate on this comment? I'm not sure I follow...

File: pkg/server/api/svid/v1/server_test.go [hunk 17]
```diff
@@ -2,35 +2,201 @@ package svid_test
...
+	}
+
+	var certs []*x509.Certificate
+	for _, cert := range s.certChain {
+		certs = append(certs, &x509.Certificate{Raw: cert})
+	}
+	return &api.X509SVID{ID: s.id, CertChain: certs, ExpiresAt: s.expiresAt}, nil
+}
+
+type testConfig struct {
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:170]: nit: 
```suggestion
type serverTest struct {
```

File: pkg/server/api/svid/v1/server_test.go [hunk 18]
```diff
@@ -2,35 +2,201 @@ package svid_test
...
+	svidClient svidpb.SVIDClient
+	service    *FakeService
+	logHook    *test.Hook
+	closeConn  func()
+}
+
+func (c *testConfig) close() {
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:177]: Even though the type is unexported, the export status of functions is a good indication about what methods on the unexported struct are "public" and not internal details to other code in the package.
```suggestion
func (c *testConfig) Close() {
```

File: pkg/server/api/svid/v1/server_test.go [hunk 19]
```diff
@@ -2,35 +2,201 @@ package svid_test
...
+		certs = append(certs, &x509.Certificate{Raw: cert})
+	}
+	return &api.X509SVID{ID: s.id, CertChain: certs, ExpiresAt: s.expiresAt}, nil
+}
+
+type testConfig struct {
+	svidClient svidpb.SVIDClient
+	service    *FakeService
+	logHook    *test.Hook
+	closeConn  func()
```
Comment (azdagron) [inline pkg/server/api/svid/v1/server_test.go:174]: nit: the `done` callback both shuts down the server and closes the client connection. `closeConn` only conveys half that. 
```suggestion
	done  func()
```

File: pkg/server/api/svid/v1/service.go [hunk 20]
```diff
@@ -25,14 +30,65 @@ type Service interface {
...
-	return &service{}
+	return &service{
+		ServerCA:    config.ServerCA,
+		TrustDomain: config.TrustDomain,
+	}
 }
 
 type service struct {
+	ServerCA    ca.ServerCA
+	TrustDomain spiffeid.TrustDomain
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:41]: These fields should probably not be exported? Also, we can use more terse names for the unexported fields if you want without sacrificing readability, IMO :)
```suggestion
	ca    ca.ServerCA
	td spiffeid.TrustDomain
```

File: pkg/server/api/svid/v1/service.go [hunk 21]
```diff
@@ -25,14 +30,65 @@ type Service interface {
...
+	log := rpccontext.Logger(ctx)
+
+	if err := csr.CheckSignature(); err != nil {
+		log.WithError(err).Error("Invalid CSR: signature verify failed")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: signature verify failed")
+	}
+
+	if len(csr.URIs) != 1 {
+		log.Error("Invalid CSR: a valid URI is required")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: a valid URI is required")
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:54]:
```suggestion
		log.Error("Invalid CSR: URI SAN is required")
		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: URI SAN is required")
```

File: pkg/server/api/svid/v1/service.go [hunk 22]
```diff
@@ -25,14 +30,65 @@ type Service interface {
...
+	}
+
+	spiffeID, err := spiffeid.FromURI(csr.URIs[0])
+	if err != nil {
+		log.WithError(err).Error("Invalid CSR: a valid SPIFFE ID is expected")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: a valid SPIFFE ID is expected: %v", err)
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:60]:
```suggestion
		log.WithError(err).Error("Invalid CSR: URI SAN is not a valid SPIFFE ID")
		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: URI SAN is not a valid SPIFFE ID: %v", err)
```

File: pkg/server/api/svid/v1/service.go [hunk 23]
```diff
@@ -25,14 +30,65 @@ type Service interface {
...
 func (s *service) MintX509SVID(ctx context.Context, csr *x509.CertificateRequest, ttl time.Duration) (*api.X509SVID, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+	log := rpccontext.Logger(ctx)
+
+	if err := csr.CheckSignature(); err != nil {
+		log.WithError(err).Error("Invalid CSR: signature verify failed")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: signature verify failed")
+	}
+
+	if len(csr.URIs) != 1 {
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:52]: Maybe something like:
```
switch {
case len(csr.URIs) == 0:
    // return error that URI SAN is required
case len(csr.URIs) > 1:
    // return error that only one URI SAN is expected
}
```

File: pkg/server/api/svid/v1/service.go [hunk 24]
```diff
@@ -25,14 +30,65 @@ type Service interface {
...
+
+	spiffeID, err := spiffeid.FromURI(csr.URIs[0])
+	if err != nil {
+		log.WithError(err).Error("Invalid CSR: a valid SPIFFE ID is expected")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: a valid SPIFFE ID is expected: %v", err)
+	}
+
+	if !spiffeID.MemberOf(s.TrustDomain) {
+		log.Error("Invalid CSR: SPIFFE ID is not member of the server trust domain")
+		return nil, status.Error(codes.InvalidArgument, "invalid CSR: SPIFFE ID is not member of the server trust domain")
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service.go:65]:
```suggestion
		log.Error("Invalid CSR: SPIFFE ID is not a member of the server trust domain")
		return nil, status.Error(codes.InvalidArgument, "invalid CSR: SPIFFE ID is not a member of the server trust domain")
```

File: pkg/server/api/svid/v1/service_test.go [hunk 25]
```diff
@@ -0,0 +1,246 @@
...
+	"github.com/spiffe/spire/test/clock"
+	"github.com/spiffe/spire/test/fakes/fakeserverca"
+	"github.com/spiffe/spire/test/spiretest"
+	"github.com/spiffe/spire/test/testkey"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+)
+
+func TestServiceMintX509SVID(t *testing.T) {
+	clock := clock.NewMock(t)
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service_test.go:26]: nit: Unless you need to control the clock, you can just omit this option and the `fakeserverca` will create it's own (that you can get at via the `Clock()` method like you are below.

File: pkg/server/api/svid/v1/service_test.go [hunk 26]
```diff
@@ -0,0 +1,246 @@
...
+				return createCsr(t, template, key)
+			},
+			dns:       []string{"dns1", "dns2"},
+			expiredAt: now.Add(time.Minute),
+			spiffeID:  spiffeID,
+			subject:   "CN=dns1,O=ORG,C=US+C=EN",
+			ttl:       time.Minute,
+			x509CA:    x509CA,
+		}, {
+			name: "default values",
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service_test.go:80]: are there other "default values" being tested here other than TTL? If not, maybe this test should be changed to test that we can override the TTL with the request and the rest of the tests can use the default. The benefit is that this would be the only test case that has to set the `ttl` field on the test case instead of the other way around.

File: pkg/server/api/svid/v1/service_test.go [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+		code      codes.Code
+		createCsr func(tb testing.TB) *x509.CertificateRequest
+		dns       []string
+		err       string
+		expiredAt time.Time
+		msg       string
+		spiffeID  spiffeid.ID
+		subject   string
+		ttl       time.Duration
+		x509CA    *ca.X509CA
```
Comment (azdagron) [inline pkg/server/api/svid/v1/service_test.go:57]: this seems to only be used to cause a minting failure from the CA. Maybe instead of requiring this to be set on most test cases, we change it to something like the following:
```suggestion
		failMinting bool
```

Then only the test that needs to assert that the server CA responds to the failure needs to set the value to true in the test case. The rest of the tests can rely on the zero value.

File: pkg/server/api/svid/v1/service.go (hunk 20 above)
Comment (MarcosDY) [reply pkg/server/api/svid/v1/service.go:41]: 👍

File: pkg/server/api/svid/v1/service_test.go (hunk 26 above)
Comment (MarcosDY) [reply pkg/server/api/svid/v1/service_test.go:80]: iit is testing subject as well, and no dns is returned

File: pkg/common/x509util/dns_test.go [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+			err: "length exceeded",
+		},
+		{
+			name: "dot only dns",
+			dns:  ".",
+			err:  "label is empty",
+		},
+		{
+			name: "ending dot dns",
+			dns:  "abcd.",
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/common/x509util/dns_test.go:41]: While true for DNS in general, RFC5280, which governs the DNS SAN says that the DNS SAN ``MUST` use the "preferred name syntax" (presumably for better interop), which does not include a trailing dot. Also, many libraries have traditionally mishandled the trailing dot.

https://tools.ietf.org/html/rfc5280#section-4.2.1.6
https://tools.ietf.org/html/rfc1034#section-3.5
